### PR TITLE
Feature/spectrum hit annotation/branchof/general spectrum reference annotation

### PR DESCRIPTION
**Pull request by mwalzer**
This is the part of split rework of #1797 , closed after scope creep. 
It is based on #249 (suggested merge after). Following the introduced SpectrumMetaDataLookup changes, this introduces SpectrumAnnotator, using TheoreticalSpectrumGenerator and SpectrumAlignment to match and annotate given identifications with given peak data at a given tolerance. Which ion types to mach is controlled via the Param extension. Also some match statistics are added.
Annotation looks something like:
```
<UserParam type="string" name="matched_ions" value="y6+,b13+,y9+,b2+,a11+,y5+,y4+,y3+,a12-C1H2N2+,y10-C1H4O1S1+,c5-H2O1+,b9++,c10-H3N1++,b11+,b12+,b15+,y10+,b9+,c10-H3N1+,..."/>
<UserParam type="float" name="matched_intensity" value="100662.793632507"/>
<UserParam type="int" name="matched_ion_number" value="229"/>
<UserParam type="int" name="peak_number" value="440"/>
<UserParam type="float" name="sum_intensity" value="159149.309085846"/>
<UserParam type="float" name="NTermIonCurrentRatio" value="0.300109904676478"/>
<UserParam type="float" name="CTermIonCurrentRatio" value="0.202220592085132"/>                                <UserParam type="float" name="median_fragment_error" value="0.0183461949475543"/>
<UserParam type="float" name="IQR_fragment_error" value="0.159449305250064"/>
<UserParam type="float" name="topN_meanfragmenterror" value="0.00127883276817491"/>
...
```

### CODE REVIEW

File: src/openms/include/OpenMS/COMPARISON/SPECTRA/SpectrumAlignment.h [hunk 1]
```diff
@@ -37,8 +37,9 @@
 
 #include <OpenMS/DATASTRUCTURES/DefaultParamHandler.h>
 
-#include <OpenMS/KERNEL/MSSpectrum.h>
-#include <OpenMS/KERNEL/MSExperiment.h>
+//#include <OpenMS/KERNEL/MSSpectrum.h>
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/SPECTRA/SpectrumAlignment.h]: please remove comments if includes not needed

File: src/openms/include/OpenMS/KERNEL/MSSpectrum.h [hunk 2]
```diff
@@ -35,17 +35,15 @@
 #ifndef OPENMS_KERNEL_MSSPECTRUM_H
 #define OPENMS_KERNEL_MSSPECTRUM_H
 
-#include <OpenMS/KERNEL/StandardDeclarations.h>
 #include <OpenMS/METADATA/SpectrumSettings.h>
 #include <OpenMS/METADATA/MetaInfoDescription.h>
 #include <OpenMS/KERNEL/RangeManager.h>
 #include <OpenMS/KERNEL/ComparatorUtils.h>
 #include <OpenMS/METADATA/DataArrays.h>
+#include <OpenMS/KERNEL/StandardTypes.h>
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h]: why is this change needed?

File: src/openms/include/OpenMS/KERNEL/MSSpectrum.h [hunk 3]
```diff
@@ -70,6 +68,7 @@ namespace OpenMS
     public RangeManager<1>,
     public SpectrumSettings
   {
+     friend class MSSpectrumHelper;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h]: why is a friend introduced?

File: src/openms/include/OpenMS/KERNEL/MSSpectrum.h [hunk 4]
```diff
@@ -35,17 +35,15 @@
 #ifndef OPENMS_KERNEL_MSSPECTRUM_H
 #define OPENMS_KERNEL_MSSPECTRUM_H
 
-#include <OpenMS/KERNEL/StandardDeclarations.h>
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h]: can you comment on removing the forward declarations?

**Comment by timosachsenberg** [comment]
@mathias I would strongly suggest modifying this PR after https://github.com/OpenMS/OpenMS/pull/2472 was merged. Maybe @enetz can assist if there are any problems with the new spectrum alignment and the way we store the annotation without RichPeak.

File: src/openms/include/OpenMS/KERNEL/MSSpectrum.h (hunk 4 above)
Comment (mwalzer) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h]: Thanks pointing out, I corrected the includes changes introduced from cherrypick but obsolete after rebase to recent develop

File: src/openms/include/OpenMS/KERNEL/MSSpectrum.h (hunk 2 above)
Comment (mwalzer) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h]: Thanks pointing out, I corrected the includes changes introduced from cherrypick but obsolete after rebase to recent develop

File: src/openms/include/OpenMS/KERNEL/MSSpectrum.h (hunk 3 above)
Comment (mwalzer) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h]: Yep, so the friend relation is needed because the helper needs access to the DataArrays of a spectrum, which are protected, to clone the peaks from a PeakSpectrum into a `MSSpectrum<RichPeak1D>`

Comment (timosachsenberg) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h]: hmm design doesn't feel right to me. You have public read access to the dataarrays (e.g. getFloatDataArrays () const and mutable).

Comment (timosachsenberg) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h]: on the other hand this will go away when we remove RichPeakSpectra

File: src/openms/include/OpenMS/CHEMISTRY/SpectrumAnnotator.h [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+#include <OpenMS/METADATA/PeptideHit.h>
+#include <OpenMS/METADATA/PeptideIdentification.h>
+#include <OpenMS/KERNEL/MSSpectrumHelper.h>
+
+#include <boost/regex.hpp>
+
+namespace OpenMS
+{
+  /**
+      @brief Generates theoretical spectra with various options
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/CHEMISTRY/SpectrumAnnotator.h]: description too generic and also wrong

File: src/openms/include/OpenMS/KERNEL/MSSpectrum.h (hunk 3 above)
Comment (mwalzer) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h]: mh, the corresponding code does not look too awkward, so I'm gonna roll with you here. But I'm not sure how soon getting rid of rich peaks will happen. Functionality in this PR is really needed right now for moving forward on the QC front, so I am not eager to wait.

Comment (timosachsenberg) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h]: nearly all were removed in https://github.com/OpenMS/OpenMS/pull/2472/files only a handful left

**Comment by timosachsenberg** [comment]
https://github.com/OpenMS/OpenMS/pull/2472 is now merged

**Comment by mwalzer** [comment]
Great, so no chance for more QC in the next release

**Comment by timosachsenberg** [comment]
or alternatively - you just merge and fix your code

**Comment by mwalzer** [comment]
well, took longer than expected, but adds nice integrity test/ reassurance to the TSG changes, I didn't have to change the TOPP test idXML files

**Comment by timosachsenberg** [comment]
Thanks. MSSpectrumHelper is this class still required?

**Comment by mwalzer** [comment]
I would leave it in because #2472 did not eradicate `MSSpectrum<RichPeak1D>`, so it might be good to be able to clone from one to the other.

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+  {
+  }
+
+  void SpectrumAnnotator::annotateMatches(PeakSpectrum& spec, const PeptideHit& ph, const TheoreticalSpectrumGenerator& tg, const SpectrumAlignment& sa) const
+  {
+    PeakSpectrum theoretical_spec;
+    vector<pair<Size, Size> > al;
+    Int zmin = 1;
+    Int zmax = 2;
+    tg.getSpectrum(theoretical_spec, ph.getSequence(), zmin, min(ph.getCharge(),zmax));
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: space after ,

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+    if (!theoretical_spec.isSorted())
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: I think this is by design of getSpectrum. If you want to save time but keep it you could put this in a precondition so it is only checked in debug mode

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+            ph->setMetaValue("median_fragment_error", 0);
+            ph->setMetaValue("IQR_fragment_error", 0);
+            ph->setMetaValue("topN_meanfragmenterror", 0);
+            ph->setMetaValue("topN_MSEfragmenterror", 0);
+            ph->setMetaValue("topN_stddevfragmenterror", 0);
+          }
+          else
+          {
+            vector<double> fe(fragmenterrors);
+            std::size_t mid = fe.size()/2;
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: spaces around operators here and several places.
we usually use Size instead of size_t in OpenMS

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+              ph->setMetaValue("median_fragment_error", (right2mid+fe[mid-1])/2.0);
+            }
+            std::nth_element(fe.begin(),          fe.begin() + lq, fe.end());
+            std::nth_element(fe.begin() + lq + 1, fe.begin() + mid, fe.end());
+            std::nth_element(fe.begin() + mid + 1, fe.begin() + uq, fe.end());
+            ph->setMetaValue("IQR_fragment_error", fe[uq]-fe[lq]);
+
+            vector<double> topn_fe;
+            topn_fe.resize(fragmenterrors.size());
+            std::reverse_copy(fragmenterrors.begin(), fragmenterrors.end(), topn_fe.begin()); //possible due to the sortByIntensity
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: why possible due to sort by Intensity? maybe extend comment here?

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+
+            double sum = std::accumulate(topn_fe.begin(), topn_fe.end(), 0.0);
+            double mean = sum / topn_fe.size();
+
+            std::vector<double> diff(topn_fe.size());
+            std::transform(topn_fe.begin(), topn_fe.end(), diff.begin(),
+                           std::bind2nd(std::minus<double>(), mean));
+            double sq_sum = std::inner_product(diff.begin(), diff.end(), diff.begin(), 0.0);
+            double m_sq_sum = (sq_sum / topn_fe.size());
+            double stdev = std::sqrt(sq_sum / topn_fe.size());
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: we now have mean and stdev in Math:: so please use these

File: src/topp/IDFileConverter.cpp [hunk 11]
```diff
@@ -141,6 +147,58 @@ class TOPPIDFileConverter :
   {
   }
 
+private:
+  bool add_ionmatches_(vector<PeptideIdentification>& peptide_identifications, String filename,  double tolerance)
```
Comment (timosachsenberg) [inline src/topp/IDFileConverter.cpp]: double space before tolerance

File: src/topp/IDFileConverter.cpp [hunk 12]
```diff
@@ -324,6 +383,13 @@ class TOPPIDFileConverter :
           MascotXMLFile::initializeLookup(lookup, exp, scan_regex);
           PepXMLFile().load(in, protein_identifications,
                             peptide_identifications, mz_name, lookup);
+
+// not sure if this would work with that kind of lookup object
```
Comment (timosachsenberg) [inline src/topp/IDFileConverter.cpp]: remove commented code

File: src/topp/IDFileConverter.cpp [hunk 13]
```diff
@@ -335,6 +401,12 @@ class TOPPIDFileConverter :
         {
           SpectrumMetaDataLookup::addMissingSpectrumReferences(
             peptide_identifications, mz_file, false, !getFlag_("no_spectra_data_override"), protein_identifications);
+
+          double add_ions = getDoubleOption_("add_ionmatch_annotation");
+          if (add_ions>0)
```
Comment (timosachsenberg) [inline src/topp/IDFileConverter.cpp]: space around operators (please check globally in this PR)

Comment (mwalzer) [inline src/topp/IDFileConverter.cpp]: I think the comments are pinned on a previous commit.

Comment (mwalzer) [inline src/topp/IDFileConverter.cpp]: resolved

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp (hunk 7 above)
Comment (mwalzer) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: Mh, obviously something is fudging up. 
So, there is the `sortByPosition` before `return` in _getSpectrum_ but still
http://cdash.openms.de/testDetails.php?test=40435416&build=224971

File: src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp [hunk 14]
```diff
@@ -135,11 +137,12 @@ namespace OpenMS
     vector<pair<Size, Size> > al;
     Int zmin = 1;
     Int zmax = 2;
-    tg.getSpectrum(theoretical_spec, ph.getSequence(), zmin, min(ph.getCharge(),zmax));
-    if (!theoretical_spec.isSorted())
+    tg.getSpectrum(theoretical_spec, ph.getSequence(), zmin, min(ph.getCharge(), zmax));
+    OPENMS_PRECONDITION(!theoretical_spec.isSorted(), "TheoreticalSpectrumGenerator::getSpectrum did not yield a sorted spectrum!")
     {
       theoretical_spec.sortByPosition();
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp]: why check for the precondition and then sort anyways?

**Comment by timosachsenberg** [comment]
Looks good but could you add the missing class test for MSSpectrumHelper ?

**Comment by timosachsenberg** [comment]
PR continued in https://github.com/OpenMS/OpenMS/pull/2642
